modules/agent/pkg/controllers/bundledeployment/controller.go

```diff
@@ -159,6 +172,28 @@ func shouldRedeploy(bd *fleet.BundleDeployment) bool {
 	return *bd.Status.SyncGeneration != bd.Spec.Options.ForceSyncGeneration
 }
 
+func (h *handler) cleanupOldAgent(bd *fleet.BundleDeployment, modified []fleet.ModifiedStatus) error {
+	if !isAgent(bd) {
+		return nil
+	}
+	for _, modified := range modified {
+		if modified.Delete {
+			gvk := schema.FromAPIVersionAndKind(modified.APIVersion, modified.Kind)
+			mapping, err := h.restMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+			if err != nil {
+				return err
+			}
+
+			logrus.Infof("Removing old agent resource %s/%s, %s", modified.Namespace, modified.Name, gvk)
+			err = h.dynamic.Resource(mapping.Resource).Namespace(modified.Namespace).Delete(h.ctx, modified.Name, metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
```

Comment:
Perhaps, we should collect errors and `continue`. Then, we can return the slice of errors.

Reply:
done